Allow creating gallery image versions from storage accounts

jeremycline opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

It's possible to upload a VHD into a container in a storage account and then use that to import an image version. This adds a third way to specify an `os_disk` in the galleryimageversion plugin to support that approach.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
galleryimageversion

jeremycline committed:
Allow creating gallery image versions from storage accounts

It's possible to upload a VHD into a container in a storage account and
then use that to import an image version. This adds a third way to
specify an `os_disk` in the galleryimageversion plugin to support that
approach.

jeremycline commented on plugins/modules/azure_rm_galleryimageversion.py:
I realized after writing this that we could construct the URI by accepting two additional arguments, the `container_name` and `blob_name`, but I'm not sure if folks would prefer that two dropping the URI in there.

jeremycline commented on plugins/modules/azure_rm_galleryimageversion.py:
```diff
@@ -561,7 +586,7 @@ def exec_module(self, **kwargs):
                                 if kwargs[key]['os_disk']['source'].get('id') is not None:
                                     self.body['properties']['storageProfile']['osDiskImage']['source']['id'] = kwargs[key]['os_disk']['source'].get('id')
-                                if kwargs[key]['os_disk']['source'].get('resource_group') is not None and \
+                                elif kwargs[key]['os_disk']['source'].get('resource_group') is not None and \
```
I think this was a bug, I'm happy to pull it out into a separate commit/PR if folks would prefer that.

Fred-sun commented on plugins/modules/azure_rm_galleryimageversion.py:
```diff
@@ -63,7 +63,10 @@
+                            - Reference to os disk snapshot.
+                            - Could be resource ID.
+                            - Could be a dictionary containing I(resource_group) and I(name).
+                            - Could be a dictionary containing I(resource_group), I(storage_account), and I(uri) if the snapshot is stored as a PageBlob in a storage account container.
```
line too long

Fred-sun commented on plugins/modules/azure_rm_galleryimageversion.py:
```diff
@@ -561,7 +586,7 @@ def exec_module(self, **kwargs):
                                     self.body['properties']['storageProfile']['osDiskImage']['source']['id'] = kwargs[key]['os_disk']['source'].get('id')
-                                if kwargs[key]['os_disk']['source'].get('resource_group') is not None and \
+                                elif kwargs[key]['os_disk']['source'].get('resource_group') is not None and \
                                    kwargs[key]['os_disk']['source'].get('name') is not None:
```
continuation line over-indented for visual indent

Fred-sun commented on plugins/modules/azure_rm_galleryimageversion.py:
```diff
@@ -570,6 +595,18 @@ def exec_module(self, **kwargs):
                                                                                                                 '/providers/Microsoft.Compute/snapshots/' +
                                                                                                                 kwargs[key]['os_disk']['source'].get('name'))
+                                elif kwargs[key]['os_disk']['source'].get('uri') is not None and \
+                                  kwargs[key]['os_disk']['source'].get('resource_group') is not None and \
```
continuation line over-indented for visual indent

Fred-sun commented on plugins/modules/azure_rm_galleryimageversion.py:
```diff
@@ -570,6 +595,18 @@ def exec_module(self, **kwargs):
                                                                                                                 kwargs[key]['os_disk']['source'].get('name'))
+                                elif kwargs[key]['os_disk']['source'].get('uri') is not None and \
+                                  kwargs[key]['os_disk']['source'].get('resource_group') is not None and \
+                                  kwargs[key]['os_disk']['source'].get('storage_account') is not None:
```
continuation line over-indented for visual indent

Fred-sun commented on plugins/modules/azure_rm_galleryimageversion.py:
```diff
@@ -570,6 +595,18 @@ def exec_module(self, **kwargs):
+                                                                                                                self.subscription_id +
+                                                                                                                '/resourceGroups/' +
+                                                                                                                resource_group +
+                                                                                                                '/providers/Microsoft.Storage/storageAccounts/' +
```
line too long

jeremycline committed:
Address PR comments for style

There's no functional change.

jeremycline commented:
Hi, thanks for the review!

I attempted to address the style comments. If it's still not looking right, is there a style guide for the project I should follow, or a linter? I did my best to follow the existing style of the file.

Fred-sun commented on plugins/modules/azure_rm_galleryimageversion.py:
```diff
@@ -303,6 +307,28 @@
+        - name: East US
+          regional_replica_count: 1
+          storage_account_type: Standard_LRS
+    storage_profile:
```
Parameter definition sequence: storage_profile -> os_disk -> source -> uri

Fred-sun commented on tests/integration/targets/azure_rm_gallery/tasks/main.yml:
```diff
@@ -343,6 +343,46 @@
+            os_disk_image:
+              disk_encryption_set_id: "{{ des_results.state.id }}"
+          storage_account_type: Standard_ZRS
+    storage_profile:
```
Parameter definition sequence: storage_profile -> os_disk -> source -> uri
```e.g.

383     storage_profile:
384       os_disk:
385         source:
386           resource_group: "{{ resource_group }}"
387           storage_account: '{{ vms_output.vms[0].storage_account_name }}'
388           uri: 'https://{{ vms_output.vms[0].storage_account_name }}.blob.core.windows.net/{{ vms_output.vms[0].stor    age_container_name }}/{{ vms_output.vms[0].storage_blob_name }}'

Fred-sun replied:
@jeremycline The output variable on line 377 should be the one obtained by the VM on line 108. But lines 108 through 380 also register other output variables. So should pay attention to 120, 377, 378 this variable value acquisition! Thank you very much!

jeremycline replied:
Ah I see, it's being shadowed. My Ansible is very rusty, sorry. I _think_ I've fixed it.

Fred-sun reviewed (commented):
Update test case! Thank you very much！

jeremycline committed:
Fix example and tests based on review comments

Add the missing "source" intermediate key in the example and test case,
and ensure the VM information isn't shadowed by later steps.

jeremycline commented:
I think this should be ready for a re-review (time permitting, of course). We're using this in some automation in Fedora so I'd love it to be in the next release if possible. Thanks!

jeremycline commented:
Hi @Fred-sun,

Just wanted to check in if there's anything I can do to help move this along - we've been carrying this patch to upload Fedora images nightly for a couple months and it's been working fine, for what it's worth.

Fred-sun commented:
@jeremycline Ok, I will finish the PR review today. Thank you!

Fred-sun commented:
@jeremycline Is the location of your test case improperly arranged? It is between line 307 and 387. Should we create a new name of galleryimageversion?

Fred-sun commented:
@jeremycline In addition, after the completion of the test, please delete the newly created galleryimage, otherwise there will be  problems when deleting the gallery. Thank you!

jeremycline committed:
galleryimageversion: rename image in test created from URI

This also removes the image version at the end so the gallery can be
deleted.

jeremycline commented:
Thanks for the review!

I've given the image version a different name and ensured it's deleted with the other image version.

Fred-sun commented:
@jeremycline Thanks for your update! I will recheck  it and push for merged!
